core/node/events/stream.go

```diff
@@ -134,6 +132,79 @@ func (s *streamImpl) ApplyMiniblock(ctx context.Context, miniblock *MiniblockInf
 	return s.applyMiniblockImplNoLock(ctx, miniblock)
 }
 
+// importMiniblock imports the given miniblocks.
+func (s *streamImpl) importMiniblocks(
+	ctx context.Context,
+	miniblocks []*MiniblockInfo,
+) error {
+	if len(miniblocks) == 0 {
+		return nil
+	}
+
+	blocksToWriteToStorage := make([]*storage.LatestMiniBlock, len(miniblocks))
+	for i, miniblock := range miniblocks {
+		bytes, err := miniblock.ToBytes()
+		if err != nil {
+			return err
+		}
+
+		blocksToWriteToStorage[i] = &storage.LatestMiniBlock{
+			StreamID:      s.streamId,
+			Number:        miniblock.Num,
+			MiniBlockInfo: bytes,
+		}
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.view == nil {
+		importFromGenesis := miniblocks[0].header().MiniblockNum == 0
+		if importFromGenesis {
+			if err := s.initFromGenesis(ctx, miniblocks[0], blocksToWriteToStorage[0]); err != nil {
+				return err
+			}
+			miniblocks = miniblocks[1:]
+			blocksToWriteToStorage = blocksToWriteToStorage[1:]
+		}
+
+		if err := s.loadInternal(ctx); err != nil {
+			return err
+		}
+	}
+
+	// apply mini-blocks one by one on view, backup existing view in case
+	// applying/writing miniblocks fails rollback view.
+	var (
+		err              error
+		viewBeforeImport = s.view
```

Comment:
it's less error-prone to construct new view and then assign it to the s.view at the end: if there are more than one error returns, it's easy to forget to flip it back (in other words, it's better to have `if err != nil { return err }` without extra cleanup steps)